### ISSUE TITLE
Docker cleanup error on mounted files

### ISSUE THREAD

**Issue by hra0031** [apptainer/singularity#758]
We are experiencing an issue with Singularity docker files cleanup. After every run (random docker image) the singularity.runtime folder remain in temp and singularity ends with error:
```
WARNING [U=500,P=16752]    unlink()                                  Failed removing file: /tmp/.singularity-runtime.6PXI2t0m/centos:latest/usr/share/zoneinfo/UTC
ERROR   [U=500,P=16752]    main()                                    Could not remove directory /tmp/.singularity-runtime.6PXI2t0m: Device or resource busy
ABORT   [U=500,P=16752]    main()                                    Retval = 255
```

I have to modify the source code of cleanup to not stop on the first error and continue. Then I can see what files can't be removed. On CentOS 7 (inside singularity) it is:
```
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/usr/share/zoneinfo/UTC
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/tmp
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/etc/passwd
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/etc/group
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/etc/hosts
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/etc/resolv.conf
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/dev
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/sys
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/var/tmp
/tmp/.singularity-runtime.Pb5XAzR6/centos:latest/proc
```
After some investigation I found, that every of this files is a mountpoint which is mounted on start of singularity:
```
/dev/sda31 on /usr/share/zoneinfo/UTC type ext3 (rw,nosuid,relatime,errors=continue,barrier=1,data=ordered)
/dev/sda31 on /etc/hosts type ext3 (rw,nosuid,relatime,errors=continue,barrier=1,data=ordered)
proc on /proc type proc (rw,relatime)
sysfs on /sys type sysfs (rw,nosuid,relatime)
devtmpfs on /dev type devtmpfs (rw,nosuid,relatime,size=65944376k,nr_inodes=16486094,mode=755)
(rw,relatime,cpuset,noprefix,release_agent=/sbin/cpuset_release_agent)
/dev/sda33 on /tmp type ext3 (rw,nosuid,relatime,errors=continue,barrier=1,data=ordered)
/dev/sda31 on /var/tmp type ext3 (rw,nosuid,relatime,errors=continue,barrier=1,data=ordered)
tmpfs on /etc/passwd type tmpfs (rw,nosuid,relatime,size=16384k)
tmpfs on /etc/group type tmpfs (rw,nosuid,relatime,size=16384k)
/dev/sda31 on /etc/resolv.conf type ext3 (rw,nosuid,relatime,errors=continue,barrier=1,data=ordered)
```
I try to remove files as user/root while singularity cleanup is still running (I have to add some delay to code) This ends with the same error. `lsof` and `fuser` can't see anything on this files. When the cleanup process close, files can be removed with `rm` without problem.
As next thing, I try to unmount mounted files directly in C code with escalated privileges before removing and I get the same error: `Device or resource busy`.
For now, I'm at end of the road and I don't know what to try as next thing.

### Version of Singularity:

Centos/RHEL 6.7/6.9
Singularity 2.3/develop

### Steps to reproduce behavior

```
[vagrant@centos6]$ singularity --debug shell -e docker://centos:latest
...
Singularity centos:latest:~/singularity> exit
VERBOSE [U=500,P=16752]    main()                                    Cleaning directory: /tmp/.singularity-runtime.6PXI2t0m
DEBUG   [U=500,P=16752]    s_rmdir()                                 Removing directory: /tmp/.singularity-runtime.6PXI2t0m
WARNING [U=500,P=16752]    unlink()                                  Failed removing file: /tmp/.singularity-runtime.6PXI2t0m/centos:latest/usr/share/zoneinfo/UTC
ERROR   [U=500,P=16752]    main()                                    Could not remove directory /tmp/.singularity-runtime.6PXI2t0m: Device or resource busy
ABORT   [U=500,P=16752]    main()                                    Retval = 255
```

**Comment by gmkurtzer** [comment]
Hi @hra0031,

I am not able to reproduce this, but I've made some changes to cleanupd in the master GitHub branch. Can you test that when you get a chance?

Thanks!

**Comment by hra0031** [comment]
I just tried it and it's working. Where was the problem? With the location of the lock file? Now it's working on all our OS. Thank you for great work!

**Comment by gmkurtzer** [comment]
Excellent! Honestly, I don't know exactly what fixed it, but I'm just happy that the latest master is working as expected! Whew!

Thanks!

**Comment by bravemaster3** [comment]
@hra0031 & @gmkurtzer 

Please, I am experiencing a problem, and was wondering if it is maybe related.
**Here is my error**

"INFO:    Cleanup error: while unmounting /var/apptainer/mnt/session/underlay/code directory: device or resource busy, while unmounting /var/apptainer/mnt/session/final directory: device or resource busy
Killed
"
**Which occurs when I run opendronemap container with the following command**
singularity run --nv --bind /path_to/datasets:/datasets odm_gpu.sif --dsm --orthophoto-resolution 3.0 --project-path /datasets

The code works, binds the /datasets on the image, and runs several things before returning this error when moving from one stage to another.
I don't know much about singularity, and I am just using it for processing drone images on a HPC.
If you have any suggestions for me, I would be grateful.

**apptainer version: apptainer version 1.1.0-1.el8**

Thanks!

**Comment by bravemaster3** [comment]
I just tried again, with apptainer instead of singularity with the same code, and it worked.
Either the HPC memories were full due to several concurrent use by others, or using  apptainer run solved the issue. will check later.
Thanks.

**Comment by DrDaveD** [comment]
That INFO message was an intermittent warning, not fatal.  I think it must have been unrelated to your "Killed" message.  The message was introduced in 1.1.0, and fixed in 1.1.5 with apptainer/apptainer#974.

**Comment by bravemaster3** [comment]
> That INFO message was an intermittent warning, not fatal. I think it must have been unrelated to your "Killed" message. The message was introduced in 1.1.0, and fixed in 1.1.5 with [apptainer/apptainer#974](https://github.com/apptainer/apptainer/issues/974).

Okay. Good to know! Thank you :)
